resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCredentialRenewer.scala

```diff
@@ -0,0 +1,150 @@
+        runnable.run()
+      } else {
+        logInfo(s"Scheduling login from keytab in $remainingTime millis.")
+        credentialRenewerThread.schedule(runnable, remainingTime, TimeUnit.MILLISECONDS)
+      }
+    }
+
+    val credentialRenewerRunnable =
+      new Runnable {
+        override def run(): Unit = {
+          try {
+            val creds = getRenewedDelegationTokens(conf)
+            broadcastDelegationTokens(creds)
+          } catch {
+            case e: Exception =>
+              // Log the error and try to write new tokens back in an hour
+              logWarning("Couldn't broadcast tokens, trying agin in 20 seconds", e)
+              credentialRenewerThread.schedule(this, 20, TimeUnit.SECONDS)
+              return
+          }
+          scheduleRenewal(this)
+        }
+      }
+    scheduleRenewal(credentialRenewerRunnable)
+  }
+
+  private def getRenewedDelegationTokens(conf: SparkConf): Array[Byte] = {
+    logInfo(s"Attempting to login with ${conf.get("spark.yarn.principal", null)}")
+    // Get new delegation tokens by logging in with a new UGI
+    // inspired by AMCredentialRenewer.scala:L174
+    val ugi = if (mode == "keytab") {
+      UserGroupInformation.loginUserFromKeytabAndReturnUGI(principal, secretFile)
+    } else {
+      UserGroupInformation.getUGIFromTicketCache(secretFile, principal)
+    }
+    val tempCreds = ugi.getCredentials
+    val hadoopConf = SparkHadoopUtil.get.newConfiguration(conf)
+    var nextRenewalTime = Long.MaxValue
+    ugi.doAs(new PrivilegedExceptionAction[Void] {
+      override def run(): Void = {
+        nextRenewalTime = tokenManager.obtainDelegationTokens(hadoopConf, tempCreds)
+        null
+      }
+    })
+
+    val currTime = System.currentTimeMillis()
+    timeOfNextRenewal = if (nextRenewalTime <= currTime) {
+      logWarning(s"Next credential renewal time ($nextRenewalTime) is earlier than " +
+        s"current time ($currTime), which is unexpected, please check your credential renewal " +
+        "related configurations in the target services.")
+      currTime
+    } else {
+      val rt = 0.75 * (nextRenewalTime - currTime)
+      (currTime + rt).toLong
+    }
+    logInfo(s"Time of next renewal is $timeOfNextRenewal")
+
+    // Add the temp credentials back to the original ones.
+    UserGroupInformation.getCurrentUser.addCredentials(tempCreds)
+    SparkHadoopUtil.get.serialize(tempCreds)
+  }
+
+  private def broadcastDelegationTokens(tokens: Array[Byte]): Unit = {
+    // send token to existing executors
+    logInfo("Sending new tokens to all executors")
+    de.send(UpdateDelegationTokens(tokens))
+  }
+}
+
+object MesosCredentialRenewer extends Logging {
+  def getTokenRenewalInterval(bytes: Array[Byte], conf: SparkConf): Long = {
+    val hadoopConf = SparkHadoopUtil.get.newConfiguration(conf)
+    val creds = SparkHadoopUtil.get.deserialize(bytes)
+    val intervals = creds.getAllTokens.asScala.flatMap { t =>
+      Try {
```

Comment:
t -> token
This method does not return an interval, it just returns the new expiration time. 
Compare with: https://github.com/apache/spark/blob/b9ab791a9efb0dc165ba283c91acf831fa6be5d8/core/src/main/scala/org/apache/spark/deploy/security/HadoopFSDelegationTokenProvider.scala#L102